Test-suite fails to build on OS X

Note I'm using Nix on OS X.

```
$ cabal build
Preprocessing library for inline-c-cpp-0.2.2.1..
Building library for inline-c-cpp-0.2.2.1..
ignoring (possibly broken) abi-depends field for packages
Preprocessing test suite 'tests' for inline-c-cpp-0.2.2.1..
Building test suite 'tests' for inline-c-cpp-0.2.2.1..
[1 of 1] Compiling Main             ( test/tests.hs, dist/build/tests/tests-tmp/Main.o )
clang-5.0: warning: argument unused during compilation: '-nopie' [-Wunused-command-line-argument]

/var/folders/8m/tqlhlzrd4lsbb8r3j9zw7wsw0000gn/T/ghc13912_0/ghc_4.cpp:127:8: error:
     warning: generalized initializer lists are a C++11 extension [-Wc++11-extensions]
    |
127 | return {};
    |        ^
return {};
       ^~

/var/folders/8m/tqlhlzrd4lsbb8r3j9zw7wsw0000gn/T/ghc13912_0/ghc_4.cpp:127:8: error:
     error: scalar initializer cannot be empty
    |
127 | return {};
    |        ^
return {};
       ^~

... cut ...

12 warnings and 12 errors generated.

<no location info>: error:
    `cc' failed in phase `C Compiler'. (Exit code: 1)
```

Since the documentation of `addForeignFile` mentions:

    the flags passed as part of -optc will be also applied to the C compiler invocation 
    that will compile them.

I tried forcing `clang` to use C++ (`-xc++`) and the C++11 standard (`-std=c++11`):

```
$ cabal build --ghc-options="-optc-xc++ -optc-std=c++11"
Preprocessing library for inline-c-cpp-0.2.2.1..
Building library for inline-c-cpp-0.2.2.1..
ignoring (possibly broken) abi-depends field for packages
Preprocessing test suite 'tests' for inline-c-cpp-0.2.2.1..
Building test suite 'tests' for inline-c-cpp-0.2.2.1..
Linking dist/build/tests/tests ...
clang-5.0: warning: argument unused during compilation: '-nopie' [-Wunused-command-line-argument]

In file included from /var/folders/8m/tqlhlzrd4lsbb8r3j9zw7wsw0000gn/T/ghc19845_0/ghc_1.c:1:0: error:


In file included from /nix/store/nc6wvwy7skyz2nsh7hyhgshg5a0y56j5-ghc-8.4.3/lib/ghc-8.4.3/include/Rts.h:29:0: error:


/nix/store/nc6wvwy7skyz2nsh7hyhgshg5a0y56j5-ghc-8.4.3/lib/ghc-8.4.3/include/Stg.h:29:3: error:
     error: __STDC_VERSION__ does not advertise C99 or later
   |
29 | # error __STDC_VERSION__ does not advertise C99 or later
   |   ^
# error __STDC_VERSION__ does not advertise C99 or later
  ^
1 error generated.
`cc' failed in phase `C Compiler'. (Exit code: 1)
```

Note that [this condition](https://github.com/ghc/ghc/blob/master/includes/Stg.h#L28) is likely failing because `__STDC_VERSION__` is undefined when compiling in C++ mode.

Maybe the condition needs  to support C++ as in:

```C++
#if !(__STDC_VERSION__ >= 199901L || __cplusplus >= 201103L)
# error __STDC_VERSION__ does not advertise C99 or later nor does __cplusplus advertise C++11 or later
#endif
```

Of course I can just define `__STDC_VERSION__` myself but that leads to all other kinds of errors:

```
$ cabal build --ghc-options="-optc-xc++ -optc-std=c++11 -optc-D__STDC_VERSION__=199901L"
Preprocessing library for inline-c-cpp-0.2.2.1..
Building library for inline-c-cpp-0.2.2.1..
ignoring (possibly broken) abi-depends field for packages
Preprocessing test suite 'tests' for inline-c-cpp-0.2.2.1..
Building test suite 'tests' for inline-c-cpp-0.2.2.1..
Linking dist/build/tests/tests ...
clang-5.0: warning: argument unused during compilation: '-nopie' [-Wunused-command-line-argument]

In file included from /var/folders/8m/tqlhlzrd4lsbb8r3j9zw7wsw0000gn/T/ghc20020_0/ghc_1.c:1:0: error:


In file included from /nix/store/nc6wvwy7skyz2nsh7hyhgshg5a0y56j5-ghc-8.4.3/lib/ghc-8.4.3/include/Rts.h:29:0: error:


In file included from /nix/store/nc6wvwy7skyz2nsh7hyhgshg5a0y56j5-ghc-8.4.3/lib/ghc-8.4.3/include/Stg.h:222:0: error:


In file included from /nix/store/nc6wvwy7skyz2nsh7hyhgshg5a0y56j5-ghc-8.4.3/lib/ghc-8.4.3/include/stg/Types.h:37:0: error:


In file included from /nix/store/3kiymx2x7isvafzlqxv1rxn8arwvj8bw-clang-wrapper-5.0.2/resource-root/include/inttypes.h:30:0: error:


/nix/store/nbcf9myn3rs5i0ic303wjmgr3jx1a934-Libsystem-osx-10.11.6/include/inttypes.h:251:35: error:
     error: expected ')'
    |
251 | strtoimax(const char * __restrict __nptr,
    |                                   ^
strtoimax(const char * __restrict __nptr,
...
```

So I guess `-optc-xc++` is not really the way to go.

--- a/.travis.yml
+++ b/.travis.yml
@@ -11,19 +11,33 @@ addons:
     - libgmp-dev
     - libgsl0-dev
     - libcairo2-dev
+  homebrew:
+    packages:
+    - cairo
+    - gsl
 
-env:
-- ARGS="--flag inline-c:gsl-example"
-- ARGS="--stack-yaml stack-lts-12.14.yaml --flag inline-c:gsl-example"
-# gtk2hs-buildtools is not present in nightly a bit of a pain to install,
-# skip it for now
-- ARGS="--stack-yaml stack-nightly-2018-10-24.yaml"
+
+matrix:
+  include:
+   - env: ARGS="--flag inline-c:gsl-example"
+   - env: ARGS="--stack-yaml stack-lts-12.14.yaml --flag inline-c:gsl-example"
+   # gtk2hs-buildtools is not present in nightly a bit of a pain to install,
+   # skip it for now
+   - env: ARGS="--stack-yaml stack-nightly-2018-10-24.yaml"
+   - env: ARGS="--flag inline-c:gsl-example"
+     os: osx
 
 before_install:
 # Download and unpack the stack executable
 - mkdir -p ~/.local/bin
 - export PATH=$HOME/.local/bin:$PATH
-- travis_retry curl -L https://www.stackage.org/stack/linux-x86_64 | tar xz --wildcards --strip-components=1 -C ~/.local/bin '*/stack'
+- |
+  if [ `uname` = "Darwin" ]
+  then
+    travis_retry curl --insecure -L https://get.haskellstack.org/stable/osx-x86_64.tar.gz | tar xz --strip-components=1 --include '*/stack' -C ~/.local/bin
+  else
+    travis_retry curl -L https://get.haskellstack.org/stable/linux-x86_64.tar.gz | tar xz --wildcards --strip-components=1 -C ~/.local/bin '*/stack'
+  fi
 
 # This line does all of the work: installs GHC if necessary, build the library,
 # executables, and test suites, and runs the test suites. --no-terminal works

--- a/inline-c-cpp/inline-c-cpp.cabal
+++ b/inline-c-cpp/inline-c-cpp.cabal
@@ -20,14 +20,17 @@ source-repository head
 library
   exposed-modules:     Language.C.Inline.Cpp
                        Language.C.Inline.Cpp.Exceptions
-  ghc-options:         -Wall
   build-depends:       base >=4.7 && <5
                      , inline-c >= 0.6.1.0
                      , template-haskell
                      , safe-exceptions
   hs-source-dirs:      src
   default-language:    Haskell2010
-  cc-options:          -Wall -Werror
+  ghc-options:         -Wall -optc-std=c++11
+  if os(darwin)
+    -- avoid https://gitlab.haskell.org/ghc/ghc/issues/11829
+    ld-options:        -Wl,-keep_dwarf_unwind 
+    ghc-options:       -pgmc=clang++
 
 test-suite tests
   type:                exitcode-stdio-1.0
@@ -39,5 +42,9 @@ test-suite tests
                      , safe-exceptions
                      , hspec
   default-language:    Haskell2010
+  ghc-options:
+    -optc-std=c++11
+  if os(darwin)
+    ghc-options: -pgmc=clang++
   extra-libraries:     stdc++
-  cc-options:          -Wall -Werror
+  cc-options:          -Wall -Werror -std=c++11

--- a/stack.yaml
+++ b/stack.yaml
@@ -1,5 +1,9 @@
-resolver: lts-11.22
+resolver: lts-13.19
 packages:
 - inline-c
 - inline-c-cpp
 - sample-cabal-project
+extra-deps:
+- Chart-cairo-1.9.1
+- cairo-0.13.6.0
+- gtk2hs-buildtools-0.13.5.0

